static/app/components/charts/treeCoverageSunburstChart.tsx

```diff
@@ -0,0 +1,229 @@
+ * The ideal data structure for the sunburst chart coming from the backend is
+ * as follows:
+ *
+ * {
+ *   name: string;
+ *   fullPath: string;
+ *   value: 1 | undefined - value should be only for file nodes, directory nodes should not have a value
+ *   coverage: number - coverage should be for both file and directory nodes
+ *   children: <recursive structure>[] | undefined - only dir nodes should have the children field;
+ * }
+ */
+export interface TreeCoverageSunburstData {
+  children: TreeCoverageSunburstData[];
+  coverage: number;
+  fullPath: string;
+  name: string;
+  value?: number;
+}
+
+interface ProcessedTreeCoverageSunburstData extends TreeCoverageSunburstData {
+  depth?: number;
+  itemStyle?: SunburstSeriesOption['itemStyle'];
+  type?: 'dir' | 'file';
+}
+
+interface TreeCoverageSunburstChartProps {
+  data: TreeCoverageSunburstData;
+  autoHeightResize?: boolean;
+}
+
+export function TreeCoverageSunburstChart({
+  data,
+  autoHeightResize = false,
+}: TreeCoverageSunburstChartProps) {
+  const theme = useTheme();
+
+  const [rootNode] = useState(() => {
+    // we want to copy this prop value as we're inline editing it's content
+    const tree = structuredClone(data);
+    const nodeMap = new Map<string, TreeCoverageSunburstData>();
+    const edges = new Map<any, TreeCoverageSunburstData>();
+    // Store node and its depth
+    const queue: Array<[ProcessedTreeCoverageSunburstData, number]> = [[tree, 0]];
+
+    while (queue.length > 0) {
+      const queueEntry = queue.shift();
+      if (!queueEntry) {
+        continue;
+      }
+
+      const [node, currentDepth] = queueEntry;
+      node.depth = currentDepth;
+      nodeMap.set(node.fullPath, node);
+
+      if (node.children) {
+        node.type = 'dir';
+
+        if (node.coverage < 60) {
+          node.itemStyle = {color: theme.red300};
+        } else if (node.coverage < 80) {
+          node.itemStyle = {color: theme.yellow300};
+        } else {
+          node.itemStyle = {color: theme.green300};
+        }
+
+        node.children.forEach(child => {
+          queue.push([child, currentDepth + 1]);
+          edges.set(child.fullPath, node);
+        });
+      } else {
+        node.type = 'file';
+
+        if (node.coverage < 60) {
+          node.itemStyle = {color: theme.red300, opacity: 0.6};
+        } else if (node.coverage < 80) {
+          node.itemStyle = {color: theme.yellow300, opacity: 0.6};
+        } else {
+          node.itemStyle = {color: theme.green300, opacity: 0.6};
+        }
+      }
+    }
+
+    return {tree, nodeMap, edges};
+  });
+  const [renderData, setRenderData] = useState(bfsFilter(rootNode.tree, 2));
+  const [breadCrumbs, setBreadCrumbs] = useState(rootNode.tree.fullPath);
+
+  return (
+    <div>
+      <ChartContainer autoHeightResize={autoHeightResize}>
+        <ReactEchartsCore
+          echarts={echarts}
+          onEvents={{
+            mouseover: (params: Parameters<EChartMouseOverHandler>[0]) => {
```

Comment:
I've seen the useCallback pattern used in other components to organize fns up top, say `const onMouseOver = useCallback(...)` thoughts on that vs this?

Reply:
I'll probably leave this as is, as we're not running into re-rendering issues, and according to the React docs: "You should only rely on `useCallback` as a performance optimization".